dali/pipeline/operator/sequence_operator.h

```diff
@@ -0,0 +1,451 @@
+      }
+    }
+  }
+
+  virtual void ExpandOutputs(const workspace_t<Backend> &ws) {
+    for (int output_idx = 0; output_idx < ws.NumInput(); output_idx++) {
+      const auto &expand_desc = GetOutputExpandDesc(ws, output_idx);
+      auto num_expand_dims = expand_desc.NumDims();
+      if (ws.template OutputIsType<GPUBackend>(output_idx)) {
+        auto expanded_handle = ExpandOutputHelper<GPUBackend>(ws, output_idx, num_expand_dims);
+        AddOutputHelper<GPUBackend>(expanded_handle);
+      } else {
+        auto expanded_handle = ExpandOutputHelper<CPUBackend>(ws, output_idx, num_expand_dims);
+        AddOutputHelper<CPUBackend>(expanded_handle);
+      }
+    }
+  }
+
+  virtual void PostprocessOutputs(workspace_t<Backend> &ws) {
+    auto num_output = ws.NumOutput();
+    for (int output_idx = 0; output_idx < num_output; output_idx++) {
+      const auto &expand_desc = GetOutputExpandDesc(ws, output_idx);
+      auto layout_prefix = expand_desc.ExpandedLayout();
+      if (ws.template OutputIsType<GPUBackend>(output_idx)) {
+        SetOutputLayoutHelper<GPUBackend>(ws, output_idx, layout_prefix);
+      } else {
+        SetOutputLayoutHelper<CPUBackend>(ws, output_idx, layout_prefix);
+      }
+    }
+  }
+
+  virtual void CoalesceOutputShapes(std::vector<OutputDesc> &output_desc,
+                                    const workspace_t<Backend> &ws) {
+    for (size_t output_idx = 0; output_idx < output_desc.size(); output_idx++) {
+      const auto &expand_desc = GetOutputExpandDesc(ws, output_idx);
+      const auto &shape = output_desc[output_idx].shape;
+      DALI_ENFORCE(shape.num_samples() == expand_desc.NumElements(),
+                   make_string("Unexpected number of frames inferred in the operator for output ",
+                               output_idx, ". Expected ", expand_desc.NumElements(),
+                               " but the operator returned ", shape.num_samples(), "."));
+      output_desc[output_idx].shape = fold_outermost_like(shape, expand_desc.ExpandedShape());
+    }
+  }
+
+  virtual bool ProcessOutputDesc(std::vector<OutputDesc> &output_desc,
+                                 const workspace_t<Backend> &ws, bool is_inferred) {
+    if (is_inferred && expand_) {
+      CoalesceOutputShapes(output_desc, ws);
+    }
+    return is_inferred;
+  }
+
+  virtual void ExpandArgument(const workspace_t<Backend> &ws, const std::string &arg_name,
+                             const TensorVector<CPUBackend> &arg_tensor) {
+    const auto &expand_desc = GetArgExpandDesc(arg_name);
+    auto expanded_arg = ExpandArgumentLike(arg_tensor, arg_name, expand_desc);
+    auto expanded_handle = std::make_shared<TensorVector<CPUBackend>>(std::move(expanded_arg));
+    AddArgumentInputHelper(arg_name, expanded_handle);
+  }
+
+  virtual void ExpandArguments(const workspace_t<Backend> &ws) {
+    for (const auto &arg_input : ws) {
+      auto shared_tvec = arg_input.second.tvec;
```

Comment:
```suggestion
      auto &shared_tvec = arg_input.second.tvec;
```
Copying a shared pointer isn't exactly free.

Reply:
done